File: src/main/domain/algorithm/TACPP46.java
package domain.algorithm;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import domain.stock.Bought;
import domain.stock.History;
import utils.Pair;

final class TACPP46 extends Algorithm {
    /*===========================================================*/
    /*===========================================================*/
    // Private Field(s)

    private ArrayDeque<Double>  lastInputArr        = new ArrayDeque<>();
    private ArrayDeque<Double>  emaHistory          = new ArrayDeque<>();

    private Map<Bought, Double> trailingHigh        = new HashMap<>();
    private List<Bought>        markedForSelling    = new ArrayList<>();

    /*===========================================================*/
    /*===========================================================*/
    // Public Interface(s)

    @Override
    public AlgorithmOutput Run(final List<Bought> holdings, final double allocatedToke, final double currentPrice) {
        AlgorithmOutput.Buy buy  = null;
        AlgorithmOutput.Sell sell = null;

        final List<Double> ema = new ArrayList<>(emaHistory);
        final double std = utils.Math.stdDev(ema);
        final double rsi = utils.Math.rsi(ema);
        final double ma  = utils.Math.average(ema);

        final double lower_band = ma - 4.0d * std * ma;

        // Buy
        if (rsi <= 30.0d && currentPrice <= lower_band) {
            if (lastInputArr.isEmpty()) {
                lastInputArr.add(currentPrice);
            } else if (utils.Math.average(new ArrayList<>(lastInputArr)) <= currentPrice) {
                
                final double confidence = utils.Math.clamp(((1.0d - std * 100.0d) + (100.0d - rsi) / 100.0d) / 2.0d, 0.0d, 0.3d);  // changing confidence has a massive effect on returns
                final long amount = (long) (allocatedToke * confidence / currentPrice);

                if(amount != 0L) buy = new AlgorithmOutput.Buy(amount);

            } else {
                lastInputArr.add(currentPrice);
                if (lastInputArr.size() > 5) lastInputArr.poll();
            }
        } else {
            lastInputArr.clear();
        }

        final double risk = utils.Math.clamp(std * 100d, 0.05d, 0.2d); // to put it into percentages

        // Sell
        final List<Pair<Bought, Long>> toBeSold = new ArrayList<>();

        // Trailing-profit logic
        for (final var item : holdings) {
            boolean isMarked = markedForSelling.contains(item);

            // Activate trailing if gained > risk
            if (!isMarked && currentPrice > item.price * (1.0d + risk)) {
                markedForSelling.add(item);
                trailingHigh.put(item, currentPrice);
                isMarked = true;
            }

            if (isMarked) {
                double high = trailingHigh.getOrDefault(item, currentPrice);

                // Update trailing high if still rising
                if (currentPrice > high) {
                    high = currentPrice;
                    trailingHigh.put(item, high);
                }

                // Sell if price falls more than risk from peak
                if (currentPrice < high * (1.0d - risk)) {
                    toBeSold.add(new Pair<>(item, item.amount));

                    // cleanup
                    markedForSelling.remove(item);
                    trailingHigh.remove(item);
                }
            }
        }

        // Stop-loss
        for (final var item : holdings) {
            if (currentPrice < item.price * (1.0d - risk * 2.0d)) {
                toBeSold.add(new Pair<>(item, item.amount));

                // cleanup
                markedForSelling.remove(item);
                trailingHigh.remove(item);
            }
        }

        if (!toBeSold.isEmpty()) sell = new AlgorithmOutput.Sell(toBeSold);
        return new AlgorithmOutput(buy, sell);
    }

    /*===========================================================*/

    @Override
    public void UpdateHistory(final History history) {
        final double alpha = 2.0d / (emaHistory.size() + 1.0d);
        final double last = emaHistory.peekLast();

        final double newEma = alpha * history.closingPrice + (1.0d - alpha) * last;

        emaHistory.pollFirst();
        emaHistory.addLast(newEma);
    }

    /*===========================================================*/
    /*===========================================================*/
    // Constructor(s)

    /**
     * Java equivalent of C++ Init::Init_EMA(q0, q1).
     * q0: first slidingWindow prices
     * q1: next slidingWindow prices
     */
    TACPP46(final Algorithm.Init init, final List<History> emaInit) {
        super(init);

        // SlidingWindow
        final int SW = 21;
        if (emaInit.size() < 2 * SW) throw new IllegalArgumentException("Init EMA: not enough history for Initialisation");
        
        final List<History> historyQ0;
        final List<History> historyQ1;
        switch(init) {

            case Algorithm.Init.TRADING: {

                final int n = emaInit.size();
                historyQ0 = List.copyOf(emaInit.subList(n - 2 * SW, n - SW));
                historyQ1 = List.copyOf(emaInit.subList(n - SW, n));
                emaInit.subList(n - 2 * SW, n).clear();

            } break;

            case Algorithm.Init.BACKTEST: {

                historyQ0 = List.copyOf(emaInit.subList(0, SW));
                historyQ1 = List.copyOf(emaInit.subList(SW, 2 * SW));
                emaInit.subList(0, 2 * SW).clear();

            } break;

            default: throw new IllegalArgumentException("Init");

        }

        final List<Double> q0 = historyQ0.stream().map(item -> item.closingPrice).toList();
        final List<Double> q1 = historyQ1.stream().map(item -> item.closingPrice).toList();

        final double alpha = 2.0d / (q1.size() + 1.0d);
        double ema = utils.Math.average(q0); // initial Value

        for (final double price : q1) {
            ema = alpha * price + (1.0d - alpha) * ema;
            emaHistory.add(ema);
        }

        if (emaHistory.isEmpty()) throw new IllegalStateException("EMA is Empty");
    }
}

File: src/main/kotlin/domain/algorithm/TACPP46.kt
package domain.algorithm

import domain.market.security.SecurityHistory
import domain.market.security.SecurityHolding
import java.util.ArrayDeque
import java.util.Deque

//===========================================================//
/**
 * An implementation of [TradingAlgorithm].
 */
//===========================================================//

internal class TACPP46: ITradingAlgorithm {
    //===========================================================//
    //===========================================================//
    // Private Field(s)

    private val m_SlidingWindow = 21

    private val m_EmaHistory: Deque<Double>

    private val m_TrailingHigh: MutableMap<SecurityHolding, Double>
    private val m_MarkedForSelling: MutableList<SecurityHolding>

    private val m_LastInputArr: Deque<Double>

    //===========================================================//
    //===========================================================//
    // Public Method(es)

    override fun run(holdings: List<SecurityHolding>, allocatedCapital: Double, currentPrice: Double): TradingAlgorithm.Output {
        var buy: TradingAlgorithm.Output.Buy? = null
        var sell: TradingAlgorithm.Output.Sell? = null

        val ema: List<Double> = ArrayList(m_EmaHistory)
        val std: Double = domain.utils.Math.stdDev(ema)
        val rsi: Double = domain.utils.Math.rsi(ema)
        val ma: Double = domain.utils.Math.average(ema)

        val lowerBand = ma - 4.0 * std * ma

        // Buy
        if (rsi <= 30.0 && currentPrice <= lowerBand) {
            if (m_LastInputArr.isEmpty()) {
                m_LastInputArr.add(currentPrice)
            } else if (domain.utils.Math.average(ArrayList(m_LastInputArr)) <= currentPrice) {
                val confidence = Math.clamp(
                    ((1.0 - std * 100.0) + (100.0 - rsi) / 100.0) / 2.0,
                    0.0,
                    0.3
                ) // changing confidence has a massive effect on returns
                val amount = (allocatedCapital * confidence / currentPrice).toInt()

                if (amount != 0) buy = TradingAlgorithm.Output.Buy(amount)
            } else {
                m_LastInputArr.add(currentPrice)
                if (m_LastInputArr.size > 5) m_LastInputArr.poll()
            }
        } else {
            m_LastInputArr.clear()
        }

        val risk: Double = Math.clamp(std * 100.0, 0.05, 0.2) // to put it into percentages

        // Sell
        val toBeSold: MutableList<Pair<SecurityHolding, Int>> = ArrayList()

        // Trailing-profit logic
        for (item in holdings) {
            var isMarked = m_MarkedForSelling.contains(item)

            // Activate trailing if gained > risk
            if (!isMarked && currentPrice > item.entryPrice * (1.0 + risk)) {
                m_MarkedForSelling.add(item)
                m_TrailingHigh[item] = currentPrice
                isMarked = true
            }

            if (isMarked) {
                var high: Double = m_TrailingHigh.getOrDefault(item, currentPrice)

                // Update trailing high if still rising
                if (currentPrice > high) {
                    high = currentPrice
                    m_TrailingHigh[item] = high
                }

                // Sell if price falls more than risk from peak
                if (currentPrice < high * (1.0 - risk)) {
                    toBeSold.add(Pair(item, item.amount))

                    // cleanup
                    m_MarkedForSelling.remove(item)
                    m_TrailingHigh.remove(item)
                }
            }
        }

        // Stop-loss
        for (item in holdings) {
            if (currentPrice < item.entryPrice * (1.0 - risk * 2.0)) {
                toBeSold.add(Pair(item, item.amount))

                // cleanup
                m_MarkedForSelling.remove(item)
                m_TrailingHigh.remove(item)
            }
        }

        // Update State
        run {
            val alpha = 2.0 / (m_EmaHistory.size + 1.0)
            val last = m_EmaHistory.peekLast()

            val newEma = alpha * currentPrice + (1.0 - alpha) * last

            m_EmaHistory.pollFirst()
            m_EmaHistory.addLast(newEma)
        }

        if (!toBeSold.isEmpty()) sell = TradingAlgorithm.Output.Sell(toBeSold)
        return TradingAlgorithm.Output(buy, sell)
    }

    //===========================================================//
    //===========================================================//
    // Constructor(s)

    /**
     * Java equivalent of C++ Init::Init_EMA(q0, q1).
     * q0: first slidingWindow prices
     * q1: next slidingWindow prices
     */
    constructor(emaInit: List<SecurityHistory>) {
        require(emaInit.size >= 2 * m_SlidingWindow) { "Init EMA" }

        val historyQ0 = emaInit.subList(0, m_SlidingWindow).toList()
        val historyQ1 = emaInit.subList(m_SlidingWindow, 2 * m_SlidingWindow).toList()

        val q0 = historyQ0.stream().map { it.closingPrice }.toList()
        val q1 = historyQ1.stream().map { it.closingPrice }.toList()

        val alpha = 2.0 / (q1.size + 1.0)
        var ema = domain.utils.Math.average(q0) // initial Value

        for (price in q1) {
            ema = alpha * price + (1.0 - alpha) * ema
            m_EmaHistory.add(ema)
        }

        check(m_EmaHistory.size == 21) { "EMA" }
    }

    //===========================================================//

    init {
        m_EmaHistory = ArrayDeque()

        m_TrailingHigh = HashMap()
        m_MarkedForSelling = ArrayList()

        m_LastInputArr = ArrayDeque()
    }
}